### PR TITLE
[SCB-2216]breakpoints testing for fast timeout of execution

### PR DESCRIPTION
**Pull request by liubao68**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java [hunk 1]
```diff
@@ -482,4 +489,50 @@ public boolean isThirdPartyInvocation() {
...
+  private void initCurrentTimeContext() {
+    // make sure time elapsed is not null
+    this.addLocalContext(Const.CONTEXT_TIME_ELAPSED, 0L);
+
+    // not initialized
+    // 1. when first time received request
+    // 2. when first time send request
+    // initialized
+    // 1. send request in the progress processing request
+    if (this.getLocalContext(Const.CONTEXT_TIME_CURRENT) != null) {
```
Comment (wujimin) [inline core/src/main/java/org/apache/servicecomb/core/Invocation.java:533]: org.apache.servicecomb.core.invocation.InvocationStageTrace#start  
try this?

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java [hunk 2]
```diff
@@ -482,4 +489,50 @@ public boolean isThirdPartyInvocation() {
 
     return future;
   }
+
+  public void ensureInvocationNotTimeout() throws InvocationException {
```
Comment (wujimin) [inline core/src/main/java/org/apache/servicecomb/core/Invocation.java:493]: create a new handler to implement this feature?

**Review by wujimin — CHANGES_REQUESTED** [review]
use handler/HttpClientFilter/HttpServerFilter to extend this feature

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java (hunk 1 above)
Comment (liubao68) [reply core/src/main/java/org/apache/servicecomb/core/Invocation.java]: fixed

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java (hunk 2 above)
Comment (liubao68) [reply core/src/main/java/org/apache/servicecomb/core/Invocation.java:499]: Now using event to check timeout

**Comment by liubao68** [comment]
> use handler/HttpClientFilter/HttpServerFilter to extend this feature

This will make users configure difficult, I use event to decouple form code.

File: transports/transport-rest/transport-rest-client/src/main/java/org/apache/servicecomb/transport/rest/client/RestClientSenderFilter.java [hunk 3]
```diff
@@ -38,6 +38,8 @@ public String getName() {
 
   @Override
   public CompletableFuture<Response> onFilter(Invocation invocation, FilterNode nextNode) {
+    invocation.onStartSendRequest();
```
Comment (wujimin) [inline transports/transport-rest/transport-rest-client/src/main/java/org/apache/servicecomb/transport/rest/client/RestClientSenderFilter.java:41]: `invocation.getInvocationStageTrace().startSend()` is inside `RestClientSender.send`

**Comment by wujimin** [comment]
not change `org.apache.servicecomb.transport.highway.HighwayClientFilter#send`  
should change all reference of `org.apache.servicecomb.core.invocation.InvocationStageTrace#startSend` and other related StageTrace.xxx

**Review by wujimin — APPROVED** [review]

**Pull request merged**